src/main/java/com/jprobstats/ml/clustering/KMeansPlusPlusClusterer.java

```diff
@@ -0,0 +1,235 @@
+        this(k, maxIterations, new EuclideanDistance());
+    }
+
+    public KMeansPlusPlusClusterer(final int k, final int maxIterations, final DistanceMeasure measure) {
+        this(k, maxIterations, measure, new Random());
+    }
+
+    public KMeansPlusPlusClusterer(final int k, final int maxIterations, final DistanceMeasure measure,
+            final Random random) {
+        super(measure);
+        this.k = k;
+        this.maxIterations = maxIterations;
+        this.random = random;
+
+    }
+
+    @Override
+    public List<CentroidCluster<T>> cluster(Collection<T> points) throws DimensionMismatchException {
+        // sanity checks
+        MathUtils.checkNotNull(points);
+        // number of clusters has to be smaller or equal the number of data points
+        if (points.size() < k) {
+            throw new NumberIsTooSmallException("Total number of points are less than total number of clusters.");
+        }
+        // create the initial clusters
+        LOGGER.info("Number of data points :: {}", points.size());
+        LOGGER.info("Clusters size :: {}", k);
+
+        List<CentroidCluster<T>> clusters = chooseInitialCenters(points);
+        LOGGER.info("Choosen initial seeds successfully.");
+        int[] assignments = new int[points.size()];
+        assignPointsToClusters(clusters, points, assignments);
+        LOGGER.info("Initial assignment is done  successfully.");
+        final int max = maxIterations < 0 ? Integer.MAX_VALUE : maxIterations;
+        for (int i = 0; i < max; i++) {
+            List<CentroidCluster<T>> newClusters = new ArrayList<CentroidCluster<T>>();
+            for (final CentroidCluster<T> cluster : clusters) {
+                final Clusterable newCenter;
+                newCenter = centroidOf(cluster.getPoints(), cluster.getCenter().getPoint().length);
+                newClusters.add(new CentroidCluster<T>(newCenter));
+            }
+            int changes = assignPointsToClusters(newClusters, points, assignments);
+            clusters = newClusters;
+
+            // if there were no more changes in the point-to-cluster assignment
+            if (changes == 0) {
+                return clusters;
+            }
+        }
+        LOGGER.info("Max iterations and final assignment done successfully");
+        return clusters;
+    }
+
+    private Clusterable centroidOf(List<T> points, int length) {
+        double[] centroid = new double[length];
+        for (T p : points) {
+            final double[] point = p.getPoint();
+            for (int i = 0; i < point.length; i++) {
+                centroid[i] += point[i];
+            }
+        }
+        for (int i = 0; i < centroid.length; i++) {
+            centroid[i] /= points.size();
+        }
+        return new DoublePoint(centroid);
+    }
+
+    private int assignPointsToClusters(List<CentroidCluster<T>> clusters, Collection<T> points, int[] assignments) {
```

Comment:
can u add the documentation explaining the algorithm equations how you assigned it